### PR TITLE
Add unit tests for HasLicense, HasNoProhibitedPackages, and RunsAsNon…

### LINKED ISSUE #64: Add unit tests for checks HasLicense, HasNoProhibitedPackages, RunAsNonRoot
Add unit tests for the the checks that use podman run. This include checks HasLicense, HasNoProhibitedPackages, and RunAsNonRoot.

### PR DESCRIPTION
**Pull request by edreyer-1**
Added unit tests and refactored for HasLicense, HasNoProhibitedPackages, and RunsAsNonRoot checks.

Fixes #64

### CODE REVIEW

File: certification/internal/shell/has_license.go [hunk 1]
```diff
@@ -1,30 +1,43 @@
...
 
 func (p *HasLicenseCheck) Validate(image string) (bool, error) {
-	stdouterr, err := exec.Command("podman", "run", "-it", "--rm", "--entrypoint", "ls", image, "-A", "/licenses").CombinedOutput()
-	result := string(stdouterr)
+
+	return p.validate(image)
+}
+
+func (p *HasLicenseCheck) validate(image string) (bool, error) {
+	runOpts := cli.ImageRunOptions{
```
Comment (bcrochet) [inline certification/internal/shell/has_license.go:19]: Great start. But let's put the podman stuff into getDataToValidate like in base_on_ubi.

File: certification/internal/shell/has_license.go [hunk 2]
```diff
@@ -1,30 +1,43 @@
...
 	if err != nil {
-		if strings.Contains(result, "No such file or directory") || result == "" {
+
+		if strings.Contains(line, "No such file or directory") || line == "" {
 			log.Warn("license not found in the container image at /licenses")
 			return false, nil
 		}
 
-		log.Error("some error attempting to identify if /licenses container the license: ", err)
+		log.Error("some error attempting to identify if /licenses cotainer the license: ", err)
```
Comment (bcrochet) [inline certification/internal/shell/has_license.go:34]: "container"

File: certification/internal/shell/has_prohibited_packages.go [hunk 3]
```diff
@@ -2,23 +2,45 @@ package shell
...
 	"github.com/redhat-openshift-ecosystem/openshift-preflight/certification"
+	"github.com/redhat-openshift-ecosystem/openshift-preflight/cli"
 	log "github.com/sirupsen/logrus"
 )
 
 type HasNoProhibitedPackagesCheck struct{}
 
 func (p *HasNoProhibitedPackagesCheck) Validate(image string) (bool, error) {
-	stdouterr, err := exec.Command("podman", "run", "-it", "--rm", "--entrypoint", "rpm", image, "-qa", "--queryformat", "%{NAME}\n").CombinedOutput()
+	line, err := p.getDataToValidate(image)
```
Comment (bcrochet) [inline certification/internal/shell/has_prohibited_packages.go:15]: Make HasLicense look like this.

**Review by bcrochet — CHANGES_REQUESTED** [review]
A typo and a small change. Otherwise, looks good.

File: certification/internal/shell/has_license.go [hunk 4]
```diff
@@ -1,30 +1,43 @@
...
+		log.Debugf("Stdout: %s", runReport.Stdout)
+		log.Debugf("Stderr: %s", runReport.Stderr)
 		return false, err
 	}
 
-	// sanity check - in case we don't get an error, but also don't have the file.
-	if strings.Contains(result, "No such file or directory") || result == "" {
+	if strings.Contains(line, "No such file or directory") || line == "" {
 		log.Warn("license not found in the container image at /licenses")
```
Comment (samira-barouti) [inline certification/internal/shell/has_license.go:41]: could this message be part of the `HelpText`? If so, you can replace this if block with
`return  !strings.Contains(line, "No such file or directory") && line != "", nil`

File: certification/internal/shell/has_prohibited_packages.go [hunk 5]
```diff
@@ -2,23 +2,45 @@ package shell
...
+		log.Error("unable to get a list of all packages in the image, error: ", err)
+		log.Debugf("Stdout: %s", runReport.Stdout)
+		log.Debugf("Stderr: %s", runReport.Stderr)
+		return "", err
+	}
+	return runReport.Stdout, nil
+}
+
+func (p *HasNoProhibitedPackagesCheck) validate(line string) (bool, error) {
+	scanner := bufio.NewScanner(strings.NewReader(line))
```
Comment (samira-barouti) [inline certification/internal/shell/has_prohibited_packages.go:43]: IMO, we should not place the `return` statement in LoC 48. We should keep scanning the image and print all the prohibited packages, and then return at the end of the function. This helps the end users to see all the prohibited packages in one run.

File: certification/internal/shell/runs_as_nonroot.go [hunk 6]
```diff
@@ -1,37 +1,52 @@
...
+		EntryPointArgs: []string{"-u"},
+		LogLevel:       "debug",
+		Image:          image,
+	}
+	runReport, err := podmanEngine.Run(runOpts)
+	if err != nil {
+		log.Error("unable to get the id of the runtime user of this image, error: ", err)
+		log.Debugf("stdout: %s", runReport.Stdout)
+		log.Debugf("stderr: %s", runReport.Stderr)
+		return "", err
```
Comment (samira-barouti) [inline certification/internal/shell/runs_as_nonroot.go:37]: better to return `nil` rather than "".

**Review by samira-barouti — CHANGES_REQUESTED** [review]

File: certification/internal/shell/runs_as_nonroot.go (hunk 6 above)
Comment (edreyer-1) [reply certification/internal/shell/runs_as_nonroot.go:37]: Since the return type is just string instead of []string, you can't return nil.

File: certification/internal/shell/has_prohibited_packages.go [hunk 7]
```diff
@@ -2,31 +2,60 @@ package shell
...
-				return false, nil
+				found = true
+				pkgs = append(pkgs, pkg)
 			}
 		}
 	}
+	if found {
+		log.Warn("found at least one prohibited package in the container image")
+		log.Warn("found the following prohibited packages: ", pkgs)
+		return false, nil
```
Comment (samira-barouti) [inline certification/internal/shell/has_prohibited_packages.go:57]: Thank you Emily for fixing this bug. 

Two more thing: for the sake of code clarity:
1) please replace variable `found` with `prohibitedPackageFound` 
2) remove this line and replace line 60 with `return !prohibitedPackageFound, nil`

File: certification/internal/shell/has_prohibited_packages.go [hunk 8]
```diff
@@ -2,31 +2,60 @@ package shell
...
 			if pkg == scanner.Text() {
-				log.Warn("found a prohibited package in the container image: ", pkg)
-				return false, nil
+				found = true
+				pkgs = append(pkgs, pkg)
 			}
 		}
 	}
+	if found {
+		log.Warn("found at least one prohibited package in the container image")
```
Comment (samira-barouti) [inline certification/internal/shell/has_prohibited_packages.go:55]: I suggest we move this outside of the if block and print the number of prohibited packages, no matter any is found or not

File: certification/internal/shell/has_license.go [hunk 9]
```diff
@@ -1,35 +1,48 @@
...
+		if ok {
+			log.Error("some error attempting to identify if /licenses container the license: ", err)
+			log.Debugf("Stdout: %s", runReport.Stdout)
+			log.Debugf("Stderr: %s", runReport.Stderr)
+			return "", err
+		}
+
+	}
+	line := runReport.Stdout
+	return line, nil
```
Comment (samira-barouti) [inline certification/internal/shell/has_license.go:40]: you can combine these two lines; no need to introduce new var if it is not being reused.

File: certification/internal/shell/has_license.go [hunk 10]
```diff
@@ -47,7 +60,7 @@ func (p *HasLicenseCheck) Metadata() certification.Metadata {
 
 func (p *HasLicenseCheck) Help() certification.HelpText {
 	return certification.HelpText{
-		Message:    "Container images must include terms and conditions applicable to the software including open source licensing information.",
+		Message:    "Container images must include terms and conditions applicable to the software including open source licensing information. license not found in the container image at /licenses",
```
Comment (samira-barouti) [inline certification/internal/shell/has_license.go:63]: I feel this should be sth like:
`The license must be at /licenses`

**Review by samira-barouti — APPROVED** [review]
Two little nits... It is good to merge once they are addressed

File: certification/internal/shell/has_license_test.go [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+				podmanEngine = fakeEngine
+			})
+			It("should pass Validate", func() {
+				ok, err := HasLicense.Validate("dummy/image")
+				Expect(err).ToNot(HaveOccurred())
+				Expect(ok).To(BeTrue())
+			})
+		})
+		Context("When license is not found", func() {
+			BeforeEach(func() {
```
Comment (bcrochet) [inline certification/internal/shell/has_license_test.go:33]: Please add a test when the directory exists, but it empty (basically, stdout would be "")

**Review by samira-barouti — APPROVED** [review]
/lgtm

File: certification/internal/shell/has_prohibited_packages.go [hunk 12]
```diff
@@ -2,33 +2,61 @@ package shell
...
+		log.Debugf("Stdout: %s", runReport.Stdout)
+		log.Debugf("Stderr: %s", runReport.Stderr)
+		return "", err
+	}
+	return runReport.Stdout, nil
+}
+
+func (p *HasNoProhibitedPackagesCheck) validate(line string) (bool, error) {
+	scanner := bufio.NewScanner(strings.NewReader(line))
+	prohibitedPackageFound := false
```
Comment (samira-barouti) [inline certification/internal/shell/has_prohibited_packages.go:44]: Giving this a second thought, we can get rid of this variable and use `pkgs` to figure out if the check passes or fails. It would be great to rename `pkgs` to `prohibitedPackages` or anything a bit descriptive.

File: certification/internal/shell/runs_as_nonroot.go [hunk 13]
```diff
@@ -1,37 +1,52 @@
...
 	// The output we get from the exec.Command includes returns
-	stdoutString := strings.TrimSpace(out.String())
+	stdoutString := strings.TrimSpace(line)
 	uid, err := strconv.Atoi(stdoutString)
 	if err != nil {
 		log.Error("unable to determine the runtime user id of the image")
-		log.Debug("expected a value that could be converted to an integer, and got: ", out.String())
+		log.Debug("expected a value that could be converted to an integer, and got: ", stdoutString)
 		return false, err
 	}
```
Comment (samira-barouti) [inline certification/internal/shell/runs_as_nonroot.go:52]: it would be great ro replace lines 55-59 with `return uid != 0, nil`

**Review by samira-barouti — COMMENTED** [review]
Nits...

**Review by samira-barouti — APPROVED** [review]
/lgtm

**Review by bcrochet — APPROVED** [review]
/lgtm
